File: utils/bin2c.zig
const std = @import("std");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);

    if (args.len != 2) {
        std.debug.print("usage: bin2c <input>\n", .{});
        std.process.exit(1);
    }

    const data = try std.fs.cwd().readFileAlloc(
        args[1],
        allocator,
        std.Io.Limit.limited(100 * 1024 * 1024),
    );
    defer allocator.free(data);

    var stdout_buf: [65536]u8 = undefined;
    var stdout_writer = std.fs.File.stdout().writerStreaming(&stdout_buf);
    const stdout = &stdout_writer.interface;

    try stdout.writeAll("unsigned char game_main_mrb[] = {\n");
    for (data, 0..) |byte, i| {
        if (i % 12 == 0) try stdout.writeAll("  ");
        try stdout.print("0x{x:0>2}", .{byte});
        if (i < data.len - 1) try stdout.writeAll(", ");
        if ((i + 1) % 12 == 0) try stdout.writeAll("\n");
    }
    try stdout.print(
        "\n}};\nunsigned int game_main_mrb_len = {};\n",
        .{data.len},
    );

    try stdout.flush();
}

File: utils/bin2c.c
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stdout, "usage: bin2c <input>\n");
        return 1;
    }

    FILE *f = fopen(argv[1], "rb");
    if (!f) {
        fprintf(stderr, "error: cannot open file '%s'\n", argv[1]);
        return 1;
    }

    fseek(f, 0, SEEK_END);
    long size = ftell(f);
    rewind(f);

    if (size > 100 * 1024 * 1024) {
        fprintf(stderr, "error: file too large\n");
        fclose(f);
        return 1;
    }

    unsigned char *data = malloc(size);
    if (!data) {
        fprintf(stderr, "error: out of memory\n");
        fclose(f);
        return 1;
    }

    if ((long)fread(data, 1, size, f) != size) {
        fprintf(stderr, "error: failed to read file\n");
        free(data);
        fclose(f);
        return 1;
    }
    fclose(f);

    printf("unsigned char game_main_mrb[] = {\n");
    for (long i = 0; i < size; i++) {
        if (i % 12 == 0) printf(" ");
        printf("0x%02x", data[i]);
        if (i < size - 1) printf(", ");
        if ((i + 1) % 12 == 0) printf("\n");
    }
    printf("\n};\nunsigned int game_main_mrb_len = %ld;\n", (long)size);

    free(data);
    return 0;
}
